In [2]:
import numpy as np
def test_accuracy(pred, label):

    with open(pred,'r') as f:
        pred_ls = f.readlines()
    with open(label,'r') as f:
        label_ls = f.readlines()

    def extract_label(l):
        l = l.decode('utf8')
        pos = l.split(u'EOS')[1].split(u'NOP')[0].strip().split(' ')
        length = int(l.split(u'NOE')[1].split(u'NOR')[0].strip())
        rhyme = l.split(u'NOP')[1].split(u'NOE')[0].strip()
        return pos, length, rhyme
    def extract_pred(l):
        l = l.decode('utf8')
        pos = l.split(',')[0].strip().split(' ')
        length = int(l.split(',')[1])
        rhyme = l.split(',')[2].strip()
        return pos, length, rhyme
    def list_acc(l1, l2):
        buf = []
        if len(l1) > len(l2):
            tmp = l2
            l2 = l1
            l1 = tmp
        for i in range(len(l1)):
            e1 = l1[i]
            e2 = l2[i]
            buf.append(1 if e1 == e2 else 0)
        return float(sum(buf)) / len(l2)

    pos_acc = []
    len_acc = []
    rhyme_acc = [] 
    for pred_l,label_l in zip(pred_ls, label_ls):
        p_pos, p_len, p_rhyme = extract_pred(pred_l)
        l_pos, l_len, l_rhyme = extract_label(label_l)

        pos_acc.append(list_acc(p_pos, l_pos))
        len_acc.append(1 if p_len == l_len else 0)
        rhyme_acc.append(1 if p_rhyme == l_rhyme else 0)
    print 'Accuracy,  POS : %.4f, len : %.4f, rhyme : %.4f' % (np.mean(pos_acc), np.mean(len_acc), np.mean(rhyme_acc))
    
    

In [4]:
pred = '../linear.csv'
label = '../hw3_1/all/test.csv'
test_accuracy(pred, label)

# pred = '../without.csv'
# label = '../hw3_1/all/test.csv'
# test_accuracy(pred, label)


Accuracy,  POS : 0.6780, len : 0.9688, rhyme : 0.5700


### light
Accuracy,  POS : 0.8563, len : 0.9409, rhyme : 0.5464


### linear with pos
Accuracy,  POS : 0.6780, len : 0.9688, rhyme : 0.5700
